heat/core/statistics.py

```diff
@@ -626,77 +627,94 @@ def maximum(x1, x2, out=None):
+                    "input DNDarrays must be distributed along the same axis, currently x1.split is {}, x2.split is {}".format(
+                        splits[0], splits[1]
+                    )
+                )
+            else:
+                split = splits[split_not_none]
+                comm = comms[split_not_none]
     else:
         split = None
+        comm = comms[0]
 
-    # locally: apply torch.max(x1, x2)
-    output_lshape = stride_tricks.broadcast_shape(x1.lshape, x2.lshape)
-    lresult = factories.empty(output_lshape, dtype=x1.dtype)
-    lresult._DNDarray__array = torch.max(x1._DNDarray__array, x2._DNDarray__array)
-    lresult._DNDarray__dtype = types.promote_types(x1.dtype, x2.dtype)
-    lresult._DNDarray__split = split
-    if x1.split is not None or x2.split is not None:
-        if x1.comm.is_distributed():  # assuming x1.comm = x2.comm
-            output_gshape = stride_tricks.broadcast_shape(x1.gshape, x2.gshape)
-            result = factories.empty(output_gshape, dtype=x1.dtype)
-            x1.comm.Allgather(lresult, result)
-            # TODO: adopt Allgatherv() as soon as it is fixed, Issue #233
-            result._DNDarray__dtype = lresult._DNDarray__dtype
-            result._DNDarray__split = split
-
-            if out is not None:
-                if out.shape != output_gshape:
-                    raise ValueError(
-                        "Expecting output buffer of shape {}, got {}".format(
-                            output_gshape, out.shape
-                        )
-                    )
-                out._DNDarray__array = result._DNDarray__array
-                out._DNDarray__dtype = result._DNDarray__dtype
-                out._DNDarray__split = split
-                out._DNDarray__device = x1.device
-                out._DNDarray__comm = x1.comm
-
-                return out
-            return result
-
+    # sanitize out
     if out is not None:
-        if out.shape != output_lshape:
+        if not isinstance(out, dndarray.DNDarray):
+            raise TypeError("expected `out` to be None or a DNDarray, but was {}".format(type(out)))
+        if out.gshape != output_gshape:
             raise ValueError(
-                "Expecting output buffer of shape {}, got {}".format(output_lshape, out.shape)
+                "Expecting output buffer of shape {}, got {}".format(output_gshape, out.shape)
+            )
+        if out.split is not split:
+            raise ValueError(
+                "Split axis of output buffer is inconsistent with split semantics (see documentation)."
             )
-        out._DNDarray__array = lresult._DNDarray__array
-        out._DNDarray__dtype = lresult._DNDarray__dtype
-        out._DNDarray__split = split
-        out._DNDarray__device = x1.device
-        out._DNDarray__comm = x1.comm
 
-    return lresult
+    # calculate process-local element-wise minimum
+    t_dtype = torch.promote_types(t_x[0].dtype, t_x[1].dtype)
```

Comment:
i think that we should use the heat promote types. we currently use the same rules as torch for type promotion but if we ever decide to differ from that this would cause issues. it should be the same heat function just with heat dtypes

Reply:
those are two torch tensors for a torch operation, it didn't occur to me to go through the heat dtype promotion and then back. This about type promotion is a discussion we need to have again I'm afraid.